### ISSUE TITLE
Browser does not show any PG tables when public is not default schema

### ISSUE THREAD

**Issue by timlinux** [qgis/QGIS#34465]
Using QGIS 3.11 master nightly build for mac from lutra build downloaded 10 Feb 2012

To replicate make the default schema search path point to something other than public:

```sql
-- next line will mean QGIS wont see any tables, it needs the public schema to be default
ALTER USER foo SET search_path = <schema other than public>;
```

<img width="312" alt="image" src="https://user-images.githubusercontent.com/178003/74482726-f9c4ff00-4eac-11ea-9d38-58ab25e3f3ac.png">

Changing the search path back to public resolves it:

```sql
ALTER USER foo SET search_path = public;
```
<img width="313" alt="image" src="https://user-images.githubusercontent.com/178003/74483027-840d6300-4ead-11ea-8fea-4a23f65db487.png">

**Comment by jgrocha** [comment]
Confirmed here. Even if I set the search path to include public, it fails.
```sql
ALTER DATABASE geotuga SET search_path TO caminha, public;
```

The query that fails is:
```sql
SELECT l.f_table_name,l.f_table_schema,l.f_geometry_column,upper(l.type),l.srid,l.coord_dimension,c.relkind,obj_description(c.oid),array_agg(a.attname), count(CASE WHEN t.typname IN ('geometry','geography') THEN 1 ELSE NULL END) , 1 FROM geometry_columns l,pg_class c,pg_namespace n,pg_attribute a,pg_type t WHERE c.relname=l.f_table_name AND l.f_table_schema=n.nspname AND NOT a.attisdropped AND a.attrelid=c.oid AND a.atttypid=t.oid AND a.attnum>0 AND n.oid=c.relnamespace AND has_schema_privilege(n.nspname,'usage') AND has_table_privilege(c.oid,'select') AND l.f_table_schema='caminha' GROUP BY 1,2,3,4,5,6,7,c.oid,11 UNION SELECT l.f_table_name,l.f_table_schema,l.f_geography_column,upper(l.type),l.srid,2,c.relkind,obj_description(c.oid),array_agg(a.attname), count(CASE WHEN t.typname IN ('geometry','geography') THEN 1 ELSE NULL END) , 1 FROM geography_columns l,pg_class c,pg_namespace n,pg_attribute a,pg_type t WHERE c.relname=l.f_table_name AND l.f_table_schema=n.nspname AND NOT a.attisdropped AND a.attrelid=c.oid AND a.atttypid=t.oid AND a.attnum>0 AND n.oid=c.relnamespace AND has_schema_privilege(n.nspname,'usage') AND has_table_privilege(c.oid,'select') AND l.f_table_schema='caminha' GROUP BY 1,2,3,4,5,6,7,c.oid,11 ORDER BY 2,1,3 returned 7 [ERROR: relation "geometry_columns" does not exist
             LINE 1: ...try','geography') THEN 1 ELSE NULL END) , 1 FROM geometry_c...
```

The bug is not difficult to solve.

In the previous query, if both `geometry_columns` and `geography_columns` are written with the prefix `public`, it works.

**Comment by DelazJ** [comment]
> Using QGIS 3.11 master nightly build for mac from lutra build downloaded **10 Feb 2012**

:thinking: (Back to the future? :smiley: )

**Comment by gioman** [comment]
Is this a regression?

**Comment by timlinux** [comment]
> Is this a regression?

I'm not sure - first time I tested this scenario.

**Comment by timlinux** [comment]
> I'm not sure if we should change all our code. What to you think @timlinux ? I did the changes on my local development environment, but I not sure if everything will work.

@jgrocha Yeah I don't want to break everyone else's QGIS so my edge use case works. Maybe a nice resolution would be to message the user e.g. when testing the database connection saying 'we can't read your public schema, you may not be able to use this connection effectively'.

Note that you can also move postgis to another schema it seems: https://gis.stackexchange.com/a/270522/44968

**Comment by gioman** [comment]
@timlinux it affects also 3.10 but not 3.4 so is a regression. Maybe something related to some recent change by @strk ?

**Comment by jgrocha** [comment]
There is no explicit requirement, but **QGIS requires that Postgis and QGIS supporting tables are on the search path**. Maybe we should add this requirement to the documentation.

1. This is not a regression. We always expected that Postgis functions and QGIS tables were on the search path.
2. Hard coding the `public` prefix on all QGIS supporting tables and Postgis functions limits the Postgis installation to the `public` schema. Right now, Postgis can be installed in any other schema, if included in the search path.

This issue is a kind of feature request, to make QGIS able to work even if Postgis and QGIS supporting tables are not on the search path. This requires that the user configures the schemas where QGIS should use as prefix.

### Feedback already exists

The user already has a warning when connection to the database, if Postgis is not available (on the search path). As you can see on the screenshot, there is a "**WARNING    No PostGIS support in the database**" when the user connects to the database.

![Screenshot from 2020-02-18 11-38-20](https://user-images.githubusercontent.com/163681/74733031-63ddfb00-5243-11ea-9a1d-2c6a767228a8.png)

We can add some more feedback to the user (maybe a modal dialogue), if Postgis is not available on the search path. This is easy to implement and informs the user that he must have Postgis and QGIS supporting tables on the search path, _but the warning already exists_.

### Summary

In my opinion, this issue can be closed. If it really makes sense to support Postgis on a schema not in the search path, this should became a feature request.

**Comment by rouault** [comment]
For reference we add a similar request in the OGR PG driver recently, and I added autodetection of the schema where PostGIS is, and then altering the search_path of the current connection to add it to it : https://github.com/OSGeo/gdal/commit/251a98a00ba1174a6e2a19aac14c101992d599f2

**Comment by elpaso** [comment]
Should we implement this in QGIS too? @rouault patch in GDAL seems straightforward to port but this looks more a feature request to me than a "High Priority" bug.

**Comment by gioman** [comment]
> than a "High Priority" bug

@elpaso it is a regression (apparently it worked as expected before) then is "automatically" labelled as a high priority.

**Comment by jgrocha** [comment]
> > than a "High Priority" bug
> 
> @elpaso it is a regression (apparently it worked as expected before) then is "automatically" labelled as a high priority.

I don't think this is a regression. We always expected postgis on the search path.

To support this, we should also support qgis_projects, layer_styles, etc on other schemas than public.

GDAL is more batch/task oriented. Changing the search path in GDAL probably does not have side effects. Changing the search path in QGIS may have impact, if we use different connections, with probably different search paths.

In my opinion, it is not easy peasy.

**Comment by elpaso** [comment]
@gioman ok, reading the comments I had the impression that it was not a regression, and you removed the tag: https://github.com/qgis/QGIS/issues/34465#event-3046817823

**Comment by elpaso** [comment]
IMO this isn't even a "bug".

**Comment by gioman** [comment]
> @gioman ok, reading the comments I had the impression that it was not a regression, and you removed the tag: [#34465 (comment)](https://github.com/qgis/QGIS/issues/34465#event-3046817823)

@elpaso you are right, I just scrolled too quickly.
